utils/weights.py

```diff
@@ -1,17 +1,174 @@
+    over- or under staffing.
+    """
+
+    return{
+        # Weight of fairness aspects. Rest, contracted hours and preferences should be treated as hours.
+        "rest": 20,
+        "contracted hours": 2,
+        "partial weekends": 8,
+        "isolated working days": 10,
+        "isolated off days": 10,
+        "consecutive days": 12,
+        "preferences": 1.5,
+
+        # Weight of least favored employee. Use interval [0, 1].
+        "lowest fairness score": 0.5,
+
+        # Scaling excess and deficit of demand
+        "excess demand deviation factor": 1.0,
+        "deficit demand deviation factor": 1.0,
+    }
+
+
+# Weighting of Shift Design Model
+
+def get_shift_design_weights():
     return {
-        "rest": 5,
-        "contracted hours": 1,
-        "partial weekends": 1,
-        "isolated working days": 1,
-        "isolated off days": 1,
-        "consecutive days": 1,
-        "backward rotation": 1,
-        "preferences": 1,
-        "lowest fairness score": 1,
         "demand_deviation": 5,
+        "shift_duration": 15,
+
+        # Scaling excess and deficit of demand
+        "excess demand deviation factor": 1.0,
+        "deficit demand deviation factor": 1.0,
+
     }
 
 
-def get_shift_design_weights():
-    return {"demand_deviation": 5, "shift_duration": 15}
+def get_weights(time_set, staff):
+    """
+    The reference weight of demand deviation is added. Then the weights are scaled to hours, which is the reference
+    time step. Lastly, the weights are scaled relatively to each other and to the problem specific data.
+    """
+
+    weights = set_weights()
+    weights = correct_weights(weights)
+    weights = add_demand_deviation_weight(weights)
+    weights = scale_weights_to_hours(weights, time_set)
+    weights = scale_weights_relatively(weights, staff)
+    weights = scale_up_weights(weights, staff)
+
+    return weights
+
+
+def correct_weights(weights):
+    """
+    Weekly rest contributes positively to the fairness score (if additional hours are granted), different from the
+    other fairness aspects. Thus it is necessary to correct the weight, so that it actually represents the intention
+    defined in set_weights.
+
+    NOTE:   Preferences are also contributing positively to the fairness score, but these respective weights are
+            corrected naturally in the generate_preferences-function.
+    """
+
+    weights["rest"] = 1 / weights["rest"]
+
+    return weights
+
+
+def add_demand_deviation_weight(weights):
+    """
+    Weight of demand deviation is not an input value and is thus treated in a separate function. The weight is set
+    to 1, serving as the reference value for remaining weights. This value should not be changed.
+    """
+
+    weights["demand deviation"] = 1
+
+    return weights
+
+
+def scale_weights_to_hours(weights, time_set):
+    """
+    Demand deviation, contracted hours and preferences needs to be scaled to hours, as this is the reference time step.
```

Comment:
Wouldn't this convert _from_ hours?

Reply:
What do you mean? The point is: If weight is set to 4 for demand deviation per hour (bad example, but the most intuitively) and the time step is 0.25, the true weight of demand deviation per hour would be 4 * 4, since every time period of 0.25 hour would get the weight 4. Therefore, the weight is scaled based on the time step - in this case reduced to 1.

Reply:
If my understanding is correct:
1. The user sets `weight_contracted_hours` to 4 demand deviation / hour 
2. You have to scale this to make sure that the objective is not affected by the size of `time_step`. 
3. You therefore multiple `weight_contracted_hours` with `time_step`, resulting in: 

```python
weight_contracted_hours = 4 demand deviation / hour * 0.25 hour = 1 demand devation per time_step
```

So you are scaling _from_ hours, and _towards_ `time_step`. If that is the case a better name would be `scale_to_time_step`. (As you are within the file `weights.py` it is not strictly necessary to include _weights_ in your function names)